src/lib/pubkey/classic_mceliece/cmce_field_ordering.cpp

```diff
@@ -0,0 +1,326 @@
+/*
+ * Classic McEliece Field Ordering Generation
+ * Based on the public domain reference implementation by the designers
+ * (https://classic.mceliece.org/impl.html - released in Oct 2022 for NISTPQC-R4)
+ *
+ * (C) 2023 Jack Lloyd
+ *     2023,2024 Fabian Albert, Amos Treiber - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ **/
+#include <botan/internal/cmce_field_ordering.h>
+
+#include <botan/cmce.h>
+#include <botan/mem_ops.h>
+#include <botan/internal/loadstor.h>
+
+#include <bitset>
+#include <numeric>
+#include <utility>
+#include <vector>
+
+namespace Botan {
+
+namespace CMCE_CT {
+
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+   requires(sizeof(T1) <= 8 && sizeof(T2) <= 8)
+void cond_swap_pair(CT::Mask<uint64_t> cond_mask, std::pair<T1, T2>& a, std::pair<T1, T2>& b) {
+   cond_mask.conditional_swap(a.first, b.first);
+   cond_mask.conditional_swap(a.second, b.second);
+}
+
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+void compare_and_swap_pair(std::span<std::pair<T1, T2>> a, size_t i, size_t k, size_t l) {
+   static_assert(sizeof(T1) <= sizeof(uint64_t) && sizeof(T2) <= sizeof(uint64_t),
+                 "Types T1 and T2 must be at most 64 bits wide");
+   if((i & k) == 0) {  // i and k do not depend on secret data
+      auto swap_required_mask = CT::Mask<uint64_t>::is_lt(a[l].first, a[i].first);
+      cond_swap_pair(swap_required_mask, a[i], a[l]);
+   } else {
+      auto swap_required_mask = CT::Mask<uint64_t>::is_gt(a[l].first, a[i].first);
+      cond_swap_pair(swap_required_mask, a[i], a[l]);
+   }
+}
+
+// Sorts a vector of pairs after the first element
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+void bitonic_sort_pair(std::span<std::pair<T1, T2>> a) {
+   const size_t n = a.size();
+   BOTAN_ARG_CHECK(is_power_of_2(n), "Input vector size must be a power of 2");
+
+   for(size_t k = 2; k <= n; k *= 2) {
+      for(size_t j = k / 2; j > 0; j /= 2) {
+         for(size_t i = 0; i < n; ++i) {
+            const size_t l = i ^ j;
+            if(l > i) {
+               compare_and_swap_pair(a, i, k, l);
+            }
+         }
+      }
+   }
+}
+
+template <std::unsigned_integral T>
+T min(const T& a, const T& b) {
+   auto mask = CT::Mask<T>::is_lt(a, b);
+   return mask.select(a, b);
+}
+
+}  // namespace CMCE_CT
+
+namespace {
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+std::vector<std::pair<T1, T2>> zip(const secure_vector<T1>& vec_1, const secure_vector<T2>& vec_2) {
```

Comment:
Maybe we did it to avoid forcing the caller to type hints like `zip(std::span(vec_1), std::span(vec_2))`. I admit using spans is more idiomatic. I'll change it.